### DISCUSSION TITLE
Live Value Updates

### DISCUSSION THREAD

**Discussion by thebells1111** [Podcastindex-org/podcast-namespace#547] [Ideas]
Here's a link to a repo that has a functional socket.io server and delivers the expected payload to a client.

[https://github.com/thebells1111/live-updates](https://github.com/thebells1111/live-updates)

For now, I have the live socket.io server url in the `pocast:liveItem` tag as an attribute called `liveValueLink`, but that will probably change as we flesh this out. However, for the time being, create a socket.io connection on the clinet side that connects with the server found in `liveValueLink`, then listen for a message from the server called `remoteValue`. The expected payload will look like this for anyone wanting to listen to a live session and display the contents in their own app. None of my apps are currently broadcasting lines 3 and 4, but I have them there for anticipation of polls.

```js
{
  "title": "This is the title for this block.",
  "image": "https://cdn.masto.host/podcastindexsocial/accounts/avatars/000/015/466/original/1dcbcde7619f90a7.png",
  "line": [
    "this is line 1",
    "this is line 2",
    "this is line 3",
    "this is line 4"
  ],
  "link": {
    "text": "This is the text for the link",
    "url": "https://podcastindex.social"
  },
  "description": "this would be an area for something like show notes",
  "value": {
    "type": "lightning",
    "method": "keysend",
    "destinations": [
       { "name": "The Split Kit",
        "address": "030a58b8653d32b99200a2334cfe913e51dc7d155aa0116c176657a4f1722677a3",
        "customKey": "696969",
        "customValue": "boPNspwDdt7axih5DfKs",
        "split": "5",
        "fee": "false"
      }
    ]
  },
  "type": "person",
  "feedGuid:" ("optional guid to feed if it exists,  i.e. pulling info from an existing podcast/album")
  "itemGuid:" ("optional guid to item if it exists, i.e. pulling info from an existing episode/song")
}
```

**Comment by thebells1111** [comment]
A new tag should be introduced into the namespace to make this extendable. I'm proposing

<podcast:liveValue uri="link to server" protocol="protocol use" />

I'm not aware of all the possible protocols and what other attributes they would need in the tag, but the ones I'm aware of are 'socket.io', 'websocket', and 'xmpp'.

For the time being, I'm only supporting 'socket.io' as it's the one I'm most familiar with for developing.

**Comment by mitchdowney** [comment]
@thebells1111 so @kreonjr and I have begun looking into adding `liveValueLink` and `remoteValue`, and so far the proposal makes sense, but there are 2 problems in particular that I am not sure how we address, and I think we may need to consider when defining the final spec:

1) When a listener plays a livestream that is already in the middle of a "live value update" section, how does the app receive that information? Can the app receive the `remoteValue` event on connect? If no, then can the app emit it's own event in the on connect event, something like emit `getRemoteValue`, and then it receives whatever the current `remoteValue` data is? While I think that could work in theory, I guess the websocket would need some kind of scoping handling, so when the client emits `getRemoteValue`, the websocket doesn't redundantly re-broadcast the `remoteValue` event to _all_ active clients, but only to the client that requested it.

2) How can mobile apps guarantee that the `remoteValue` is kept up to date while the app is playing in the background, or the screen is locked? My understanding is that a mobile OS may pause or disconnect the websocket connection whenever it wants, and if that happens, it is a rather serious problem, because what if the app gets stuck on a specific `remoteValue` for an hour, sending sats to that music artist, even though multiple other `remoteValue` events have been sent over that time, but missed by the app because the websocket is paused? While keeping a websocket connection active while in the app is in the background for a mobile OS may not be reliable, I believe mobile OS can more reliably make polling style requests on-demand for every X seconds of playback. Should a polling-style endpoint be recommended in the spec?

I'm not sure, but I think the answer to problem 2) might be the same as the answer to problem 1). Maybe mobile apps, while in the background, could convert into a "polling" `remoteValue` mode, which briefly connects and disconnects from the websocket every X seconds of playback, and gets whatever the current `remoteValue` is in the process.

Polling does not sound as efficient as an active websocket connection would be, so I'm not sure if polling is tenable for this purpose, but it is also only a tiny amount of metadata being requested for `remoteValue`, so maybe it could be an acceptable solution?

Anyway, I'd appreciate your and anyone else's thoughts on how we might be able to address these problems within the spec.

**Reply by mitchdowney** [reply]
I’m starting to think writing our own API service hosted on one of our servers may be the only way to get past mobile OS websocket issues. Polling requests seem reliable for media player mobile apps while in the background, but websockets are not. Might not be too difficult for us to write a service that fills the gap with polling requests while the app is in the background. So I'm not sure that 2) is a problem the spec needs to worry about, but if 1) is solvable that could be good.

**Reply by agates** [reply]
You can run a websocket connection in a background service on Android.

That's essentially what's already being used for UnifiedPush support, it's just not implemented by Podverse at that layer.

**Reply by agates** [reply]
And technically you can tell socket.io to only use HTTP long-polling, but I think that would break without being in the background anyway.

**Reply by mitchdowney** [reply]
@agates can you explain a bit more why long-polling in the background would break? What if we were long-polling on iOS every 10 seconds while in the background? iOS allows you to run functions in the background as long as the media player is playing. It's one of the exceptions to their limitations as I understand.

**Reply by agates** [reply]
> @agates can you explain a bit more why long-polling in the background would break? What if we were long-polling on iOS every 10 seconds while in the background? iOS allows you to run functions in the background as long as the media player is playing. It's one of the exceptions to their limitations as I understand.

Long polling is expected to last a lot longer than 10 seconds.  And I don't know if you have that level of control with socket.io.  I'm specifically talking about socket.io.

**Reply by mitchdowney** [reply]
I just checked the socket.io docs and one of the first things it says is it offers HTTP long-polling as a fallback. I can run some test cases using the [live-updates repo](https://github.com/thebells1111/live-updates) that @thebells1111 created and see what happens.

**Comment by Kolomona** [comment]
StevenB's implementation of podcast:liveValue in The Spit Kit appears to be working quite nicely.  Can we get this put into the spec?

I would love for more players to be using it.

**Reply by thebells1111** [reply]
Might want to explore some ways to clean it up, make it less ambiguous, make it so some of the naming matches what's in the feed better before we put anything in the spec, but I do think having live value / live chapters is a good move forward.

**Reply by agates** [reply]
Need to lower expectations of reliability on mobile experiences in documentation.  Not impossible but definitely presents a higher barrier.

I'm particularly thinking of those people that think a "license" requires the app to make v4v available.  Whatever that means ;).

**Comment by SilasVogt** [comment]
Most of this i think i understand but I have a couple of questions. 
Example, i have a podast feed with 2 splits 50/50.
With this new value info, does it override the existing value block (fully), or is it an addition and the apps are supposed to recalculate the new splits. So if i add in a guest getting 33%, it recalculates the 50/50 original splits to being 33/33 in all the podcast apps.

i also dont understand "type": "person", and what "Line" is for. just displaying live text in apps?

**Reply by thebells1111** [reply]
The idea is the valueblock sent over the websocket replaces the existing block completely, just like a valueTimeSplit (VTS) would. It's basically a VTS for a live show.

The type:person is because the original intent was this was going to be used for a panel discussion that was broadcasted live, and if the type would help with changing styling based on what was being broadcast, but I think type should probably be dropped before anything is official.

Line was to make it open ended, but I think it's too ambiguous and should be changed. I think we should decide on the most import details to include but not make it so open ended. 

I'm thinking something more like this: 

```
{
  "title": "This is the title or song for this live chapter.", 
  "author":  "This is the author or band for this live chapter.",
  "podcastName": This is the podcast name or album for this live chapter.",
  "image": "https://cdn.masto.host/podcastindexsocial/accounts/avatars/000/015/466/original/1dcbcde7619f90a7.png",
  "link": {
    "text": "This is the text for the link",
    "url": "https://podcastindex.social"
  },
  "value": {
    "type": "lightning",
    "method": "keysend",
    "destinations": [
       { "name": "The Split Kit",
        "address": "030a58b8653d32b99200a2334cfe913e51dc7d155aa0116c176657a4f1722677a3",
        "customKey": "696969",
        "customValue": "boPNspwDdt7axih5DfKs",
        "split": "5",
        "fee": "false"
      }
    ]
  }
}
```

The idea is it's a way to send a chapter and value block in real time and would be treated the same way a chapter or VTS gets treated in a recorded show.

**Comment by thebells1111** [comment]
For anyone testing, this feed [https://curiocaster.com/rss/feed.xml](https://curiocaster.com/rss/feed.xml) has a `liveItem` which also has the experimental tag in it. 
`<podcast:liveValue uri="https://curiohoster.com/event?event_id=test" protocol="socket.io"/>`
Set your app to listen to that socket using socket.io, and you'll get a new event broadcast every 30 seconds.

**Reply by mitchdowney** [reply]
Awesome! Thanks for setting this up.
